core/bin/zksync_api/src/bin/providers/proxy_utils.rs

```diff
@@ -0,0 +1,54 @@
+use std::{
+    collections::HashMap,
+    sync::Arc,
+    time::{Duration, Instant},
+};
+use tokio::sync::Mutex;
+
+use actix_web::HttpResponse;
+use serde_json::Value;
+
+pub(crate) struct ResponseCache<T> {
+    data: T,
+    last_fetched: Instant,
+}
+
+#[derive(Clone)]
+pub(crate) struct ProxyState {
+    pub cache: Arc<Mutex<HashMap<String, ResponseCache<Value>>>>,
+}
+
+pub(crate) async fn proxy_request(
+    url: &str,
+    cache: &Mutex<HashMap<String, ResponseCache<Value>>>,
+) -> HttpResponse {
+    let mut lock = cache.lock().await;
+
+    // Check cache first
+    if let Some(cached) = lock.get(url) {
+        if cached.last_fetched.elapsed() < Duration::from_secs(5) {
```

Comment:
If we don't make it configurable, I'd use a bigger number to avoid hitting the rate limit.

Reply:
this could be worth configuring I guess

Reply:
done